firewood/src/merkle/node.rs

```
@@ -402,6 +416,113 @@ impl<T> PartialEq for EncodedNode<T> {
+
+        if path_bits_len == 0 {
+            return Ok(PathWithBitsPrefix(Path(vec![])));
+        }
+
+        if path_bits_len % BITS_PER_NIBBLE != 0 {
+            return Err(serde::de::Error::custom(
+                "path length is not a multiple of 4",
+            ));
+        }
+
+        if path_bits_len > MAX_PATH_BIT_LEN {
+            return Err(serde::de::Error::custom(
+                "path bit length > {MAX_PATH_BIT_LEN}",
+            ));
+        }
+
+        // Add BITS_PER_NIBBLE so that if there are an odd number of nibbles,
+        // we need to read the final byte.
+        // e.g. if `path_bits_len` is 12 we need to read 2 bytes
+        // If we didn't add BITS_PER_NIBBLE `path_bytes_len` would be 1.
+        // Note `path_bytes_len` >= 1 because we assert `path_bits_len` > 0.
+        let path_bytes_len = (path_bits_len + BITS_PER_NIBBLE) / BITS_PER_BYTE;
+
+        let mut path = Vec::with_capacity(NIBBLES_PER_BYTE * path_bytes_len as usize);
+
+        for _ in 0..path_bytes_len {
+            let byte = seq
+                .next_element::<u8>()?
+                .ok_or(serde::de::Error::custom("not enough bytes"))?;
+
+            // Convert the byte to 2 nibbles
+            path.push(byte >> 4);
+            path.push(byte & 0b_0000_1111);
+        }
+
+        if path_bits_len % BITS_PER_BYTE != 0 {
+            // The last byte only contained one nibble.
+            let padding_nibble = path.pop().expect("path_bytes_len > 0");
+            if padding_nibble != 0 {
+                return Err(serde::de::Error::custom("padding nibble is not 0"));
+            }
+        }
+
+        Ok(PathWithBitsPrefix(Path(path)))
+    }
+}
+
+impl Serialize for PathWithBitsPrefix {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let path_length_bits: u64 = (self.0.len() as u64)
+            .checked_mul(4u64)
+            .expect("too many bits to serialize");
```

Why not propagate the error?